File: 코드/ch7/clip2/boj1074/main.py
n, r, c = map(int, input().split())
count = 0

def solve(n, r, c):
    global count
    board_size = 1 << n
    mid = board_size // 2
    if n == 0:
        return
    if r < mid and c < mid:  # 좌상단
        solve(n - 1, r, c)
    elif r < mid and c >= mid:  # 우상단
        count += mid * mid
        solve(n - 1, r, c - mid)
    elif r >= mid and c < mid:  # 좌하단
        count += mid * mid * 2
        solve(n - 1, r - mid, c)
    else:  # 우하단
        count += mid * mid * 3
        solve(n - 1, r - mid, c - mid)

solve(n, r, c)
print(count)

File: 코드/ch7/clip2/boj1074/Main.java
import java.util.Scanner;

public class Main {
    public static int n, r, c;
    public static int count = 0;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        r = sc.nextInt();
        c = sc.nextInt();
        solve(n, r ,c);
        System.out.println(count);
    }
    public static void solve(int n, int r, int c) {
        int boardSize = 1 << n; // 2^n
        int mid = boardSize / 2;
        if (n == 0) return;
        // 좌상단
        if (r < mid && c < mid) {
            solve(n - 1, r, c);
        }
        // 우상단
        else if (r < mid && c >= mid) {
            count += mid * mid;
            solve(n - 1, r, c - mid);
        }
        // 좌하단
        else if (r >= mid && c < mid) {
            count += mid * mid * 2;
            solve(n - 1, r - mid, c);
        }
        // 우하단
        else {
            count += mid * mid * 3;
            solve(n - 1, r - mid, c - mid);
        }
    }
}
